### PR TITLE
revive URI for terminal quick fixes

### LINKED ISSUE #167589: Terminal quick fix - Opener type shows [object Object]
Testing #166762

image

image

### PR DESCRIPTION
**Pull request by meganrogge**
fix #167589

Screenshot 2022-12-01 at 12 02 45 PM

### CODE REVIEW

File: .vscode/notebooks/test.py [hunk 1]
```diff
@@ -0,0 +1 @@
+print(3)
```
Comment (Tyriar) [inline .vscode/notebooks/test.py]: 👀

Comment (meganrogge) [inline .vscode/notebooks/test.py:1]:
```suggestion
```

File: src/vs/workbench/contrib/terminal/browser/xterm/quickFixAddon.ts [hunk 2]
```diff
@@ -337,34 +336,24 @@ export async function getQuickFixesForCommand(
 										});
 									},
 									tooltip: label,
+									//TODO: why is command on this object?
```
Comment (Tyriar) [inline src/vs/workbench/contrib/terminal/browser/xterm/quickFixAddon.ts:339]: These comments still needed?

File: src/vs/workbench/api/browser/mainThreadTerminalService.ts [hunk 3]
```diff
@@ -468,3 +458,10 @@ export function getOutputMatchForLines(lines: string[], outputMatcher: ITerminal
 	const match: RegExpMatchArray | null | undefined = lines.join('\n').match(outputMatcher.lineMatcher);
 	return match ? { regexMatch: match, outputLines: outputMatcher.multipleMatches ? lines : undefined } : undefined;
 }
+
+function parseQuickFix(id: string, source: string, fix: TerminalQuickFix): ITerminalQuickFix {
```
Comment (meganrogge) [inline src/vs/workbench/api/browser/mainThreadTerminalService.ts:462]: when I try to test this, the window won't open with this error:

```
0x107c95ce0). One of the two will be used. Which one is undefined.
[25077:1201/113233.427542:INFO:CONSOLE(261)] "[uncaught exception]: Error: Cannot find module 'vscode'
Require stack:
- electron/js2c/renderer_init", source: vscode-file://vscode-app/Users/meganrogge/Repos/vscode/out/bootstrap-window.js (261)
[25077:1201/113233.428003:INFO:CONSOLE(264)] "Error: Cannot find module 'vscode'
Require stack:
- electron/js2c/renderer_init
    at Module._resolveFilename (node:internal/modules/cjs/loader:940:15)
    at i._resolveFilename (node:electron/js2c/renderer_init:33:1095)
    at Module._load (node:internal/modules/cjs/loader:785:27)
    at c._load (node:electron/js2c/asar_bundle:5:13343)
    at i._load (node:electron/js2c/renderer_init:33:356)
    at Module.require (node:internal/modules/cjs/loader:1012:19)
    at require (node:internal/modules/cjs/helpers:102:18)
    at nodeRequire (v
```

when I remove lines 463-464, it works

File: src/vs/workbench/contrib/terminal/browser/xterm/quickFixAddon.ts [hunk 4]
```diff
@@ -337,34 +336,24 @@ export async function getQuickFixesForCommand(
...
-										}
-
-										if (!uri) {
-											return;
-										}
-										openerService.open(uri);
-									},
+									run: () => openerService.open(fix.uri),
 									tooltip: label,
+									//TODO: why is uri on this object?
```
Comment (meganrogge) [inline src/vs/workbench/contrib/terminal/browser/xterm/quickFixAddon.ts:356]:
```suggestion
```

File: src/vs/workbench/contrib/terminal/browser/xterm/quickFixAddon.ts [hunk 5]
```diff
@@ -337,33 +336,22 @@ export async function getQuickFixesForCommand(
 										});
 									},
 									tooltip: label,
+									//TODO: why is command on this object?
```
Comment (meganrogge) [inline src/vs/workbench/contrib/terminal/browser/xterm/quickFixAddon.ts:339]:
```suggestion
```

File: src/vs/workbench/api/browser/mainThreadTerminalService.ts (hunk 3 above)
Comment (meganrogge) [reply src/vs/workbench/api/browser/mainThreadTerminalService.ts:462]: a way around it is to instead check if `uri` in fix but not sure why the above wouldn't work

Comment (Tyriar) [reply src/vs/workbench/api/browser/mainThreadTerminalService.ts:462]: Are you importing uri from `vscode`? You need to use the core lib

File: src/vs/workbench/contrib/terminal/browser/xterm/quickFixAddon.ts [hunk 6]
```diff
@@ -344,26 +343,14 @@ export async function getQuickFixesForCommand(
 							}
 							case TerminalQuickFixType.Opener: {
 								const fix = quickFix as ITerminalQuickFixOpenerAction;
-								const label = localize('quickFix.opener', 'Open: {0}', fix.uri.toString());
+								const label = localize('quickFix.opener', 'Open: {0}', fix.uri.scheme + fix.uri.path + fix.uri.authority);
```
Comment (meganrogge) [inline src/vs/workbench/contrib/terminal/browser/xterm/quickFixAddon.ts:346]: to do - is there a function that does this?

Comment (meganrogge) [reply src/vs/workbench/contrib/terminal/browser/xterm/quickFixAddon.ts:346]: trying joinPath

Comment (Tyriar) [reply src/vs/workbench/contrib/terminal/browser/xterm/quickFixAddon.ts:346]: 👇

https://github.com/microsoft/vscode/blob/968026a46db6ad769e0d9f3a048d2c7d780de083/src/vs/workbench/services/label/common/labelService.ts#L195

Comment (meganrogge) [reply src/vs/workbench/contrib/terminal/browser/xterm/quickFixAddon.ts:346]: there isn't but should be. going to add

File: src/vs/workbench/contrib/terminal/browser/xterm/quickFixAddon.ts [hunk 7]
```diff
@@ -344,26 +343,17 @@ export async function getQuickFixesForCommand(
 							}
 							case TerminalQuickFixType.Opener: {
 								const fix = quickFix as ITerminalQuickFixOpenerAction;
-								const label = localize('quickFix.opener', 'Open: {0}', fix.uri.toString());
+								const { scheme, authority, path, query, fragment } = fix.uri;
```
Comment (meganrogge) [inline src/vs/workbench/contrib/terminal/browser/xterm/quickFixAddon.ts:346]: I wonder if we should put this somewhere general to be used by others?

Comment (meganrogge) [reply src/vs/workbench/contrib/terminal/browser/xterm/quickFixAddon.ts:346]: I noticed there's a private `_asFormatted` on `URI` which can be referenced by `URI.toString` but that says explicitly it shouldn't be used for display purposes

Comment (Tyriar) [reply src/vs/workbench/contrib/terminal/browser/xterm/quickFixAddon.ts:346]: See my comment about about the label service

Comment (meganrogge) [reply src/vs/workbench/contrib/terminal/browser/xterm/quickFixAddon.ts:346]: thanks 😄

**Review by Tyriar — DISMISSED** [review]

**Review by Tyriar — DISMISSED** [review]

File: src/vs/workbench/contrib/terminal/browser/xterm/quickFixAddon.ts [hunk 8]
```diff
@@ -344,26 +347,18 @@ export async function getQuickFixesForCommand(
 							}
 							case TerminalQuickFixType.Opener: {
 								const fix = quickFix as ITerminalQuickFixOpenerAction;
-								const label = localize('quickFix.opener', 'Open: {0}', fix.uri.toString());
+								if (!fix.uri) {
+									return;
+								}
+								const uriLabel = (fix.uri.scheme === Schemas.http || fix.uri.scheme === Schemas.https) ? encodeURI(fix.uri.toString(true)) : labelService.getUriLabel(fix.uri);
+								const label = localize('quickFix.opener', 'Open: {0}', uriLabel);
```
Comment (roblourens) [inline src/vs/workbench/contrib/terminal/browser/xterm/quickFixAddon.ts:354]: The `: ` could get lost in localization. Probably only the word `Open` should be localized

Comment (Tyriar) [reply src/vs/workbench/contrib/terminal/browser/xterm/quickFixAddon.ts:354]: @roblourens that'll probably break rtl languages

**Comment by meganrogge** [comment]
looks like I need a `pathService` and a `remoteAgentService` for the tests

**Review by Tyriar — APPROVED** [review]

**Pull request merged**